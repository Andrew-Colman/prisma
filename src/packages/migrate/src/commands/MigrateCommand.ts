import {
  arg,
  Command,
  Commands,
  format,
  HelpError,
  isError,
  unknownCommand,
  getSchemaPath,
  getCommandWithExecutor,
  isCi,
  IntrospectionEngine,
} from '@prisma/sdk'
import Debug from '@prisma/debug'
import chalk from 'chalk'
import prompt from 'prompts'
import path from 'path'
import { Migrate } from '../Migrate'
import { ensureDatabaseExists } from '../utils/ensureDatabaseExists'
import {
  EarlyAcessFlagError,
  ExperimentalFlagWithNewMigrateError,
} from '../utils/flagErrors'
import { printMigrationId } from '../utils/printMigrationId'
import { printFilesFromMigrationIds } from '../utils/printFiles'
import {
  handleUnexecutableSteps,
  handleWarnings,
} from '../utils/handleEvaluateDataloss'
import { getMigrationName } from '../utils/promptForMigrationName'
import { isOldMigrate } from '../utils/detectOldMigrate'

const debug = Debug('migrate:cmd')

/**
 * Migrate command
 */
export class MigrateCommand implements Command {
  public static new(cmds?: Commands): MigrateCommand {
    return new MigrateCommand(cmds)
  }

  private static help = format(`
    ${
      process.platform === 'win32' ? '' : chalk.bold('üèãÔ∏è  ')
    }Migrate your database with confidence

    ${chalk.bold.yellow('WARNING')} ${chalk.bold(
    "Prisma's migration functionality is currently in an experimental state.",
  )}
    ${chalk.dim(
      'When using any of the commands below you need to explicitly opt-in via the --early-access-feature flag.',
    )}

    ${chalk.bold('Usage')}

      With an existing schema.prisma:
      ${chalk.dim(
        '$',
      )} prisma migrate [command] [options] --early-access-feature

      Or specify a schema:
      ${chalk.dim(
        '$',
      )} prisma migrate [command] [options] --early-access-feature --schema=./schema.prisma

    ${chalk.bold('Options')}

      -h, --help   Display this help message
         --draft   Create a draft of a migration that can be edited locally before being applied

    ${chalk.bold('Commands')}

          up      Migrate your database up
          reset   Reset your database, all data will be lost

    ${chalk.bold('Examples')}

      Create a new migration and apply it
      ${chalk.dim('$')} prisma migrate --early-access-feature

      Reset your database
      ${chalk.dim('$')} prisma migrate reset --early-access-feature

      Create a draft of a migration
      ${chalk.dim('$')} prisma migrate --draft --early-access-feature
  `)

  private argsSpec = {
    '--help': Boolean,
    '-h': '--help',
    '--name': String,
    '-n': '--name',
    '--force': Boolean,
    '-f': '--force',
    '--draft': Boolean,
    '--schema': String,
    '--skip-generate': Boolean,
    '--experimental': Boolean,
    '--early-access-feature': Boolean,
    '--telemetry-information': String,
  }

  private constructor(private readonly cmds?: Commands) {}

  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(argv, this.argsSpec)

    if (isError(args)) {
      return this.help(args.message)
    }

    if (args['--help']) {
      return this.help()
    }

    if (args['--experimental']) {
      throw new ExperimentalFlagWithNewMigrateError()
    }

    const schemaPath = await getSchemaPath(args['--schema'])
    if (schemaPath) {
      const migrationDirPath = path.join(path.dirname(schemaPath), 'migrations')
      if (isOldMigrate(migrationDirPath)) {
        // Maybe add link to docs?
        throw Error(
          `The migrations folder contains migrations files from an older version of Prisma Migrate which is not compatible.
Delete the current migrations folder to continue and read the documentation for how to upgrade / baseline.`,
        )
      }
    }

    // running a subcommand
    if (args._[0] && this.cmds) {
      // check if we have that subcommand
      const cmd = this.cmds[args._[0]]
      if (cmd) {
        const argsForCmd = args['--early-access-feature']
          ? [
              ...args._.slice(1),
              `--early-access-feature=${args['--early-access-feature']}`,
            ]
          : args._.slice(1)
        return cmd.parse(argsForCmd)
      }

      return unknownCommand(MigrateCommand.help, args._[0])
    } else {
      // prisma migrate
      if (!args['--early-access-feature']) {
        throw new EarlyAcessFlagError()
      }

      return await this.migrate(argv)
    }
  }

  // All-in-One command
  public async migrate(argv: string[]): Promise<string> {
    const args = arg(argv, this.argsSpec)

    const schemaPath = await getSchemaPath(args['--schema'])

    if (!schemaPath) {
      const message = `Could not find a ${chalk.bold(
        'schema.prisma',
      )} file that is required for this command.\nYou can either provide it with ${chalk.greenBright(
        '--schema',
      )}, set it as \`prisma.schema\` in your package.json or put it into the default location ${chalk.greenBright(
        './prisma/schema.prisma',
      )} https://pris.ly/d/prisma-schema-location`

      // if no schema arg passed display error + help
      if (!args['--schema']) {
        throw this.help(message) as string
      }

      // else only display error
      throw new Error(message)
    }

    console.info(
      chalk.dim(
        `Prisma Schema loaded from ${path.relative(process.cwd(), schemaPath)}`,
      ),
    )

    // Automatically create the database if it doesn't exist
    const wasDbCreated = await ensureDatabaseExists('create', true, schemaPath)
    if (wasDbCreated) {
      console.info()
      console.info(wasDbCreated)
    }

    const migrate = new Migrate(schemaPath)

    if (args['--draft']) {
      let migrationName: undefined | string = undefined
      const getMigrationNameResult = await getMigrationName(args['--name'])
      if (getMigrationNameResult.userCancelled) {
        migrate.stop()
        return getMigrationNameResult.userCancelled
      } else {
        migrationName = getMigrationNameResult.name
      }

      const migrationId = await migrate.draft({
        name: migrationName,
      })
      migrate.stop()

      if (migrationId) {
        return `\nPrisma Migrate created a draft migration ${printMigrationId(
          migrationId,
        )}\n\nYou can now edit it and then apply it by running ${chalk.greenBright(
          getCommandWithExecutor('prisma migrate --early-access-feature'),
        )} again.`
      } else {
        return `\nNo migration was created. Your Prisma schema and database are already in sync.\n`
      }
    }

    const diagnoseResult = await migrate.diagnoseMigrationHistory()
    debug({ diagnoseResult })

    let isResetNeeded = false
    const hasFailedMigrations = diagnoseResult.failedMigrationNames.length > 0
    const hasEditedMigrations = diagnoseResult.editedMigrationNames.length > 0

    // if failed migration(s) or edited migration(s) print and got to reset
    if (hasFailedMigrations || hasEditedMigrations) {
      isResetNeeded = true

      if (hasFailedMigrations) {
        // migration(s), usually one, that failed to apply the the database (which may have data)
        console.info(
          `The following migration(s) failed to apply:\n- ${diagnoseResult.failedMigrationNames.join(
            '\n- ',
          )}\n`,
        )
      }

      if (hasEditedMigrations) {
        // migration(s) that were edited since they were applied to the db.
        console.info(
          `The following migration(s) were edited after they were applied:\n- ${diagnoseResult.editedMigrationNames.join(
            '\n- ',
          )}\n`,
        )
      }
    } else {
      debug({ drift: diagnoseResult.drift })
      debug({ history: diagnoseResult.history })

      if (diagnoseResult.drift) {
        // TODO it seems this condition is never true
        if (diagnoseResult.drift?.diagnostic === 'migrationFailedToApply') {
          // Migration has a problem (failed to cleanly apply to a temporary database) and
          // needs to be fixed or the database has a problem (example: incorrect version, missing extension)
          throw new Error(
            `A migration failed while applying it to the shadow database.\nFix the migration before applying it again.\n\n${diagnoseResult.drift.error})`,
          )
        } else if (diagnoseResult.drift.diagnostic === 'driftDetected') {
          if (diagnoseResult.hasMigrationsTable === false) {
            const confirmDbPushUsed = await this.confirmDbPushUsed()
            if (confirmDbPushUsed) {
              const introspectEngine = new IntrospectionEngine({
                cwd: path.dirname(schemaPath),
              })

              const introspectResult = await introspectEngine.introspect(
                migrate.getDatamodel(),
              )
              introspectEngine.stop()

              const createMigrationResult = await migrate.createMigration({
                migrationsDirectoryPath: migrate.migrationsDirectoryPath,
                migrationName: '',
                draft: true,
                prismaSchema: introspectResult.datamodel,
              })

              console.info(
                `Migration "${createMigrationResult.generatedMigrationName!}" created.`,
              )

              await migrate.markMigrationApplied({
                migrationId: createMigrationResult.generatedMigrationName!,
              })

              console.info(
                `Migration "${createMigrationResult.generatedMigrationName!}" marked applied.`,
              )

              const createMigrationOptionalResult = await migrate.createMigration(
                {
                  migrationsDirectoryPath: migrate.migrationsDirectoryPath,
                  migrationName: '',
                  draft: false,
                  prismaSchema: migrate.getDatamodel(),
                },
              )

              if (createMigrationOptionalResult.generatedMigrationName) {
                console.info(
                  `Migration "${createMigrationResult.generatedMigrationName!}" created and applied.`,
                )
              }
              migrate.stop()
              return `Operation successful.`
            } else {
              migrate.stop()
              throw Error(
                'Check init flow with introspect + SQL schema dump (TODO docs)',
              )
            }
          } else {
            // we could try to fix the drift in the future
            isResetNeeded = true
          }
        }
      }

      if (diagnoseResult.history) {
        if (diagnoseResult.history.diagnostic === 'databaseIsBehind') {
          await migrate.applyOnly()
        } else if (
          diagnoseResult.history.diagnostic === 'migrationsDirectoryIsBehind'
        ) {
          isResetNeeded = true
        } else if (diagnoseResult.history.diagnostic === 'historiesDiverge') {
          isResetNeeded = true
          // migration(s) were removed from directory since they were applied to the db.
          console.info(
            `The following migration(s) are applied to the database but missing from the local migrations directory:
- ${diagnoseResult.history.unpersistedMigrationNames.join('\n- ')}\n`,
          )
        }
      }
    }

    if (isResetNeeded) {
      if (!args['--force']) {
        // We use prompts.inject() for testing in our CI
        if (isCi() && Boolean((prompt as any)._injected?.length) === false) {
          throw Error(
            `Use the --force flag to use the migrate command in an unnattended environment like ${chalk.bold.greenBright(
              getCommandWithExecutor(
                'prisma migrate --force --early-access-feature',
              ),
            )}`,
          )
        }

        const confirmedReset = await this.confirmReset(
          await migrate.getDbInfo(),
        )
        if (!confirmedReset) {
          console.info() // empty line
          console.info('Reset cancelled.')
          process.exit(0)
        }
        await migrate.reset()
      }
    }

    const evaluateDataLossResult = await migrate.evaluateDataLoss()

    // throw error
    handleUnexecutableSteps(evaluateDataLossResult.unexecutableSteps)
    // log warnings and prompt user to continue if needed
    const userCancelled = await handleWarnings(
      evaluateDataLossResult.warnings,
      args['--force'],
    )
    if (userCancelled) {
      migrate.stop()
      return `Migration cancelled.`
    }

    let migrationName: undefined | string = undefined
    if (evaluateDataLossResult.migrationSteps.length > 0) {
      const getMigrationNameResult = await getMigrationName(args['--name'])
      if (getMigrationNameResult.userCancelled) {
        migrate.stop()
        return getMigrationNameResult.userCancelled
      } else {
        migrationName = getMigrationNameResult.name
      }
    }

    const migrationIds = await migrate.createAndApply({
      name: migrationName,
    })
    migrate.stop()

    if (migrationIds.length === 0) {
      console.info(
        `\n${chalk.green(
          'Everything is already in sync',
        )} - Prisma Migrate didn't find any schema changes or unapplied migrations.`,
      )
    } else {
      console.info(
        `\nPrisma Migrate applied the following migration(s):\n\n${chalk(
          printFilesFromMigrationIds('migrations', migrationIds, {
            'migration.sql': '',
          }),
        )}`,
      )

      // Run if not skipped
      if (!process.env.MIGRATE_SKIP_GENERATE && !args['--skip-generate']) {
        await migrate.tryToRunGenerate()
      }
    }

    return ``
  }

  private async confirmReset({
    schemaWord,
    dbType,
    dbName,
    dbLocation,
  }): Promise<boolean> {
    const confirmation = await prompt({
      type: 'confirm',
      name: 'value',
      message: `We need to reset the ${dbType} ${schemaWord} "${dbName}" at "${dbLocation}". ${chalk.red(
        'All data will be lost',
      )}.\nDo you want to continue?`,
    })

    return confirmation.value
  }

  private async confirmDbPushUsed(): Promise<boolean> {
    const confirmation = await prompt({
      type: 'confirm',
      name: 'value',
      message: `Did you use ${chalk.green('prisma db push')}?`,
    })

    return confirmation.value
  }

  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(
        `\n${chalk.bold.red(`!`)} ${error}\n${MigrateCommand.help}`,
      )
    }
    return MigrateCommand.help
  }
}
