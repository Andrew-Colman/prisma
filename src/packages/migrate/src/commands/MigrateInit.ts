import {
  arg,
  Command,
  format,
  getCommandWithExecutor,
  getSchemaPath,
  HelpError,
  isError,
} from '@prisma/sdk'
import chalk from 'chalk'
import path from 'path'
import { Migrate } from '../Migrate'
import { ensureDatabaseExists } from '../utils/ensureDatabaseExists'
import { ExperimentalFlagError } from '../utils/experimental'

export class MigrateInit implements Command {
  public static new(): MigrateInit {
    return new MigrateInit()
  }

  private static help = format(`
    Initialize migrations

    ${chalk.bold.yellow('WARNING')} ${chalk.bold(
    "Prisma's migration functionality is currently in an experimental state.",
  )}
    ${chalk.dim(
      'When using any of the commands below you need to explicitly opt-in via the --experimental flag.',
    )}

    ${chalk.bold('Usage')}

      ${chalk.dim('$')} prisma migrate init --experimental

    ${chalk.bold('Options')}

      -h, --help       Displays this help message

  `)

  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--experimental': Boolean,
      '--schema': String,
      '--telemetry-information': String,
    })

    if (isError(args)) {
      return this.help(args.message)
    }

    if (args['--help']) {
      return this.help()
    }

    if (!args['--experimental']) {
      throw new ExperimentalFlagError()
    }

    const schemaPath = await getSchemaPath(args['--schema'])

    if (!schemaPath) {
      throw new Error(
        `Could not find a ${chalk.bold(
          'schema.prisma',
        )} file that is required for this command.\nYou can either provide it with ${chalk.greenBright(
          '--schema',
        )}, set it as \`prisma.schema\` in your package.json or put it into the default location ${chalk.greenBright(
          './prisma/schema.prisma',
        )} https://pris.ly/d/prisma-schema-location`,
      )
    }

    console.info(
      chalk.dim(
        `Prisma Schema loaded from ${path.relative(process.cwd(), schemaPath)}`,
      ),
    )

    // Automatically create the database if it doesn't exist
    await ensureDatabaseExists('create', true, schemaPath)

    const migrate = new Migrate(schemaPath)
    await migrate.initialize()
    // after engine init is implemented
    // migrate.stop()

    return `\nInitialization complete.\n\nRun ${chalk.greenBright(
      getCommandWithExecutor('prisma migrate --experimental'),
    )} to create your first migration\n`
  }

  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(
        `\n${chalk.bold.red(`!`)} ${error}\n${MigrateInit.help}`,
      )
    }
    return MigrateInit.help
  }
}
