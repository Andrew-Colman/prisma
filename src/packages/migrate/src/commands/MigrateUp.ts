import {
  arg,
  Command,
  format,
  getSchemaPath,
  HelpError,
  isError,
} from '@prisma/sdk'
import chalk from 'chalk'
import path from 'path'
import { Migrate } from '../Migrate'
import { ensureDatabaseExists } from '../utils/ensureDatabaseExists'
import { ExperimentalFlagError } from '../utils/experimental'

export class MigrateUp implements Command {
  public static new(): MigrateUp {
    return new MigrateUp()
  }

  // static help template
  private static help = format(`
    Migrate your database up to a specific state.

    ${chalk.bold.yellow('WARNING')} ${chalk.bold(
    "Prisma's migration functionality is currently in an experimental state.",
  )}
    ${chalk.dim(
      'When using any of the commands below you need to explicitly opt-in via the --experimental flag.',
    )}

    ${chalk.bold('Usage')}

      ${chalk.dim('$')} prisma migrate up --experimental


    ${chalk.bold('Options')}

      -h, --help            Displays this help message
      --dev                 Checks thedatabase state and interactively ask to reset if needed before applying migrations.
      --prod, --production  Applies unapplied migrations only.

  `)

  // parse arguments
  public async parse(argv: string[]): Promise<string | Error> {
    // parse the arguments according to the spec
    const args = arg(
      argv,
      {
        '--help': Boolean,
        '-h': '--help',
        '--experimental': Boolean,
        '--dev': Boolean,
        '--production': Boolean,
        '-prod': '--production',
        '--schema': String,
        '--telemetry-information': String,
      },
      false,
    )

    if (isError(args)) {
      return this.help(args.message)
    }

    if (args['--help']) {
      return this.help()
    }

    if (!args['--experimental']) {
      throw new ExperimentalFlagError()
    }

    const schemaPath = await getSchemaPath(args['--schema'])

    if (!schemaPath) {
      throw new Error(
        `Could not find a ${chalk.bold(
          'schema.prisma',
        )} file that is required for this command.\nYou can either provide it with ${chalk.greenBright(
          '--schema',
        )}, set it as \`prisma.schema\` in your package.json or put it into the default location ${chalk.greenBright(
          './prisma/schema.prisma',
        )} https://pris.ly/d/prisma-schema-location`,
      )
    }

    console.info(
      chalk.dim(
        `Prisma Schema loaded from ${path.relative(process.cwd(), schemaPath)}`,
      ),
    )

    const migrate = new Migrate(schemaPath)

    await ensureDatabaseExists('apply', true, schemaPath)

    const result = await migrate.up()
    migrate.stop()

    return ``
  }

  // help message
  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(
        `\n${chalk.bold.red(`!`)} ${error}\n${MigrateUp.help}`,
      )
    }
    return MigrateUp.help
  }
}
