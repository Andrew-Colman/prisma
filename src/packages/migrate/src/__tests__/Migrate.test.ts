import prompt from 'prompts'
import fs from 'fs-jetpack'
import { MigrateCommand } from '../commands/MigrateCommand'
import { consoleContext, Context } from './__helpers__/context'
import { tearDownMysql } from '../utils/setupMysql'
import {
  SetupParams,
  setupPostgres,
  tearDownPostgres,
} from '../utils/setupPostgres'

const ctx = Context.new().add(consoleContext()).assemble()

process.env.GITHUB_ACTIONS = '1'
process.env.MIGRATE_SKIP_GENERATE = '1'

// TODO
// add test for isReset needed https://github.com/prisma/prisma/blob/a6c91fd6816bd33d3c44ce3bc56af06c01ddfdba/src/packages/migrate/src/Migrate.ts#L350

describe('common', () => {
  it('should fail if no schema file', async () => {
    ctx.fixture('empty')
    const result = MigrateCommand.new().parse(['--early-access-feature'])
    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`

            ! Could not find a schema.prisma file that is required for this command.
            You can either provide it with --schema, set it as \`prisma.schema\` in your package.json or put it into the default location ./prisma/schema.prisma https://pris.ly/d/prisma-schema-location

            🏋️  Migrate your database with confidence

            WARNING Prisma's migration functionality is currently in an experimental state.
            When using any of the commands below you need to explicitly opt-in via the --early-access-feature flag.

            Usage

              With an existing schema.prisma:
              $ prisma migrate [command] [options] --early-access-feature

              Or specify a schema:
              $ prisma migrate [command] [options] --early-access-feature --schema=./schema.prisma

            Options

              -h, --help   Display this help message
                 --draft   Create a draft of a migration that can be edited locally before being applied

            Commands

                  up      Migrate your database up
                  reset   Reset your database, all data will be lost

            Examples

              Create a new migration and apply it
              $ prisma migrate --early-access-feature

              Reset your database
              $ prisma migrate reset --early-access-feature

              Create a draft of a migration
              $ prisma migrate --draft --early-access-feature

          `)
  })
  it('should fail if old migrate', async () => {
    ctx.fixture('old-migrate')
    const result = MigrateCommand.new().parse(['--early-access-feature'])
    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
            The migrations folder contains migrations files from an older version of Prisma Migrate which is not compatible.
            Delete the current migrations folder to continue and read the documentation for how to upgrade / baseline.
          `)
  })
  it('should fail if no flag', async () => {
    ctx.fixture('empty')
    const result = MigrateCommand.new().parse([])
    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
            This feature is currently in Early Access. There may be bugs and it's not recommended to use it in production environments.
                  Please provide the --early-access-feature flag to use this command.
          `)
  })
  it('should fail if experimental flag', async () => {
    ctx.fixture('empty')
    const result = MigrateCommand.new().parse(['--experimental'])
    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
            Prisma Migrate was Experimental and is now in Early Access.
                  WARNING this new iteration has some breaking changes to use it it's recommended to read the documentation first and replace the --experimental flag with --early-access-feature.
          `)
  })
})

describe('sqlite', () => {
  it('first migration after init - empty.prisma', async () => {
    ctx.fixture('schema-only-sqlite')
    const result = MigrateCommand.new().parse([
      '--schema=./prisma/empty.prisma',
      '--early-access-feature',
    ])
    await expect(result).resolves.toMatchSnapshot()

    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/empty.prisma

      SQLite database dev.db created at file:dev.db


      Everything is already in sync - Prisma Migrate didn't find any schema changes or unapplied migrations.
    `)
    expect(ctx.mocked['console.log'].mock.calls).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls).toMatchSnapshot()
  })

  it('first migration after init (prompt)', async () => {
    ctx.fixture('schema-only-sqlite')
    const result = MigrateCommand.new().parse([
      '--name=first',
      '--early-access-feature',
    ])

    await expect(result).resolves.toMatchSnapshot()
    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma

      SQLite database dev.db created at file:dev.db


      Prisma Migrate applied the following migration(s):

      migrations/
        └─ 20201231000000_first/
          └─ migration.sql
    `)
    expect(ctx.mocked['console.log'].mock.calls).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls).toMatchSnapshot()
  })

  it('first migration after init (--name)', async () => {
    ctx.fixture('schema-only-sqlite')

    prompt.inject(['first'])

    const result = MigrateCommand.new().parse(['--early-access-feature'])

    await expect(result).resolves.toMatchSnapshot()
    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma

      SQLite database dev.db created at file:dev.db


      Prisma Migrate applied the following migration(s):

      migrations/
        └─ 20201231000000_first/
          └─ migration.sql
    `)
    expect(ctx.mocked['console.log'].mock.calls).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls).toMatchSnapshot()
  })

  it('first migration after init --force', async () => {
    ctx.fixture('schema-only-sqlite')
    const result = MigrateCommand.new().parse([
      '--name=first',
      '--force',
      '--early-access-feature',
    ])

    await expect(result).resolves.toMatchSnapshot()
    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma

      SQLite database dev.db created at file:dev.db


      Prisma Migrate applied the following migration(s):

      migrations/
        └─ 20201231000000_first/
          └─ migration.sql
    `)
    expect(ctx.mocked['console.log'].mock.calls.join()).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls.join()).toMatchSnapshot(``)
  })

  it('draft migration and apply (prompt)', async () => {
    ctx.fixture('schema-only-sqlite')

    prompt.inject(['some-Draft'])

    const draftResult = MigrateCommand.new().parse([
      '--draft',
      '--early-access-feature',
    ])

    await expect(draftResult).resolves.toMatchSnapshot()

    const applyResult = MigrateCommand.new().parse(['--early-access-feature'])

    await expect(applyResult).resolves.toMatchSnapshot()
    expect(
      (fs.list('prisma/migrations')?.length || 0) > 0,
    ).toMatchInlineSnapshot(`true`)
    expect(fs.exists('prisma/dev.db')).toEqual('file')
    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma

      SQLite database dev.db created at file:dev.db

      Prisma Schema loaded from prisma/schema.prisma

      Everything is already in sync - Prisma Migrate didn't find any schema changes or unapplied migrations.
    `)
    expect(ctx.mocked['console.log'].mock.calls.join()).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls.join()).toMatchSnapshot()
  })

  it('draft migration and apply (--name)', async () => {
    ctx.fixture('schema-only-sqlite')
    const draftResult = MigrateCommand.new().parse([
      '--draft',
      '--name=first',
      '--early-access-feature',
    ])

    await expect(draftResult).resolves.toMatchSnapshot()

    const applyResult = MigrateCommand.new().parse(['--early-access-feature'])

    await expect(applyResult).resolves.toMatchSnapshot()
    expect(
      (fs.list('prisma/migrations')?.length || 0) > 0,
    ).toMatchInlineSnapshot(`true`)
    expect(fs.exists('prisma/dev.db')).toEqual('file')
    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma

      SQLite database dev.db created at file:dev.db

      Prisma Schema loaded from prisma/schema.prisma

      Everything is already in sync - Prisma Migrate didn't find any schema changes or unapplied migrations.
    `)
    expect(ctx.mocked['console.log'].mock.calls.join()).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls.join()).toMatchSnapshot()
  })

  it('transition-db-push-migrate (prompt yes)', async () => {
    ctx.fixture('transition-db-push-migrate')

    prompt.inject(['y'])

    const result = MigrateCommand.new().parse(['--early-access-feature'])

    await expect(result).resolves.toMatchInlineSnapshot(`Operation successful.`)
    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma
      Migration "20201231000000_" created.
      Migration "20201231000000_" marked applied.
    `)
    expect(ctx.mocked['console.log'].mock.calls.join()).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls.join()).toMatchSnapshot(``)
  })

  it('transition-db-push-migrate (prompt no)', async () => {
    ctx.fixture('transition-db-push-migrate')

    prompt.inject([new Error()])

    const result = MigrateCommand.new().parse(['--early-access-feature'])

    await expect(result).rejects.toMatchInlineSnapshot(
      `Check init flow with introspect + SQL schema dump (TODO docs)`,
    )
    expect(
      ctx.mocked['console.info'].mock.calls.join('\n'),
    ).toMatchInlineSnapshot(`Prisma Schema loaded from prisma/schema.prisma`)
    expect(ctx.mocked['console.log'].mock.calls.join()).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls.join()).toMatchSnapshot(``)
  })
})

describe('postgresql', () => {
  const SetupParams: SetupParams = {
    connectionString:
      process.env.TEST_POSTGRES_URI_MIGRATE ||
      'postgres://prisma:prisma@localhost:5432/tests-migrate',
    dirname: './fixtures',
  }

  beforeEach(async () => {
    await setupPostgres(SetupParams).catch((e) => {
      console.error(e)
    })
  })

  afterEach(async () => {
    await tearDownPostgres(SetupParams).catch((e) => {
      console.error(e)
    })
  })

  it('schema only (--force)', async () => {
    ctx.fixture('schema-only-postgresql')

    const result = MigrateCommand.new().parse([
      '--early-access-feature',
      '--force',
    ])
    await expect(result).resolves.toThrowErrorMatchingInlineSnapshot(
      `undefined`,
    )
    expect(ctx.mocked['console.log'].mock.calls).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls).toMatchSnapshot()
    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma

      Prisma Migrate applied the following migration(s):

      migrations/
        └─ 20201231000000_/
          └─ migration.sql
    `)
  })

  it('first migration after init - empty.prisma', async () => {
    ctx.fixture('schema-only-postgresql')
    const result = MigrateCommand.new().parse([
      '--schema=./prisma/empty.prisma',
      '--early-access-feature',
    ])

    await expect(result).resolves.toMatchSnapshot()
    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/empty.prisma

      Everything is already in sync - Prisma Migrate didn't find any schema changes or unapplied migrations.
    `)
    expect(ctx.mocked['console.log'].mock.calls).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls).toMatchSnapshot()
  })

  it('first migration after init', async () => {
    ctx.fixture('schema-only-postgresql')
    const result = MigrateCommand.new().parse([
      '--name=first',
      '--early-access-feature',
    ])

    await expect(result).resolves.toMatchSnapshot()
    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma

      Prisma Migrate applied the following migration(s):

      migrations/
        └─ 20201231000000_first/
          └─ migration.sql
    `)
    expect(ctx.mocked['console.log'].mock.calls).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls).toMatchSnapshot()
  })

  it('first migration after init --force + --name', async () => {
    ctx.fixture('schema-only-postgresql')
    const result = MigrateCommand.new().parse([
      '--name=first',
      '--force',
      '--early-access-feature',
    ])

    await expect(result).resolves.toMatchSnapshot()
    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma

      Prisma Migrate applied the following migration(s):

      migrations/
        └─ 20201231000000_first/
          └─ migration.sql
    `)
    expect(ctx.mocked['console.log'].mock.calls).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls).toMatchSnapshot()
  })

  it('draft migration and apply', async () => {
    ctx.fixture('schema-only-postgresql')
    const draftResult = MigrateCommand.new().parse([
      '--draft',
      '--name=first',
      '--early-access-feature',
    ])

    await expect(draftResult).resolves.toMatchSnapshot()

    const applyResult = MigrateCommand.new().parse(['--early-access-feature'])
    await expect(applyResult).resolves.toMatchSnapshot()

    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma
      Prisma Schema loaded from prisma/schema.prisma

      Everything is already in sync - Prisma Migrate didn't find any schema changes or unapplied migrations.
    `)
    expect(ctx.mocked['console.log'].mock.calls).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls).toMatchSnapshot()
  })

  it('existingdb: first migration after init', async () => {
    ctx.fixture('schema-only-postgresql')
    const result = MigrateCommand.new().parse([
      '--name=first',
      '--early-access-feature',
    ])

    await expect(result).resolves.toMatchSnapshot()
    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma

      Prisma Migrate applied the following migration(s):

      migrations/
        └─ 20201231000000_first/
          └─ migration.sql
    `)
    expect(ctx.mocked['console.log'].mock.calls).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls).toMatchSnapshot()
  })
})

describe.skip('mysql', () => {
  const SetupParams: SetupParams = {
    connectionString: `${
      process.env.TEST_MYSQL_URI || 'mysql://prisma:prisma@localhost:3306/tests'
    }`,
    dirname: __dirname,
  }

  beforeEach(async () => {
    await tearDownMysql(SetupParams).catch((e) => {
      console.error({ e })
    })
  })

  afterAll(async () => {
    await tearDownMysql(SetupParams).catch((e) => {
      console.error({ e })
    })
  })
})
