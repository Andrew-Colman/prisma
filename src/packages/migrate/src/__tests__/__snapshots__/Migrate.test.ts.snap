// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sqlite first migration after init - empty schema 1`] = ``;

exports[`sqlite first migration after init - empty schema 3`] = `Array []`;

exports[`sqlite first migration after init - empty schema 4`] = `Array []`;
